src/helpers/comment.ts

```diff
@@ -1,42 +1,91 @@
-import * as parse5 from "parse5";
+import { MarkdownItem } from "../types";
 
 type Node = {
-  nodeName: string;
-  tagName?: string;
-  value?: string;
-  childNodes?: Node[];
+  type: MarkdownItem;
+  value: string;
+  depth?: number;
+  children: Node[];
 };
 
-const traverse = (result: Record<string, string[]>, node: Node, itemsToExclude: string[]): Record<string, string[]> => {
-  if (itemsToExclude.includes(node.nodeName)) {
-    return result;
+const traverse = (result: Record<MarkdownItem, string[]>, node: Node, itemsToExclude: string[]): Record<MarkdownItem, string[]> => {
+  if (!result[node.type]) {
+    result[node.type] = [];
   }
 
-  if (!result[node.nodeName]) {
-    result[node.nodeName] = [];
+  if (node.type === MarkdownItem.Heading) {
+    node.type = `heading${node.depth}` as MarkdownItem;
   }
 
-  result[node.nodeName].push(node.value?.trim() ?? "");
+  result[node.type].push(node.value?.trim() ?? "");
+
+  if (itemsToExclude.includes(node.type)) {
+    return result;
+  }
 
-  if (node.childNodes && node.childNodes.length > 0) {
-    node.childNodes.forEach((child) => traverse(result, child, itemsToExclude));
+  if (node.children && node.children.length > 0) {
+    node.children.forEach((child) => traverse(result, child, itemsToExclude));
   }
 
   return result;
 };
 
-export const parseComments = (comments: string[], itemsToExclude: string[]): Record<string, string[]> => {
-  const result: Record<string, string[]> = {};
+export const parseComments = async (comments: string[], itemsToExclude: string[]): Promise<Record<MarkdownItem, string[]>> => {
+  // dynamic import of mdast
+  const { fromMarkdown } = await import("mdast-util-from-markdown");
+  const { gfmFromMarkdown } = await import("mdast-util-gfm");
+  const { gfm } = await import("micromark-extension-gfm");
```

Comment:
I tried importing normally at the top but I got the error:
```
Error [ERR_REQUIRE_ESM]: require() of ES Module /Users/whilefoo/ubiquity/ubiquibot/node_modules/mdast-util-from-markdown/index.js from /Users/whilefoo/ubiquity/ubiquibot/lib/src/helpers/comment.js not supported.
    Instead change the require of index.js in /Users/whilefoo/ubiquity/ubiquibot/lib/src/helpers/comment.js to a dynamic import() which is available in all CommonJS modules.
```
After that I changed to dynamic import in the function but the error stays the same. I checked the compiled javascript and it's still `require()` - it seems every `import` is compiled to `require`.

@rndquu any ideas?

Reply:
I guess this is a related issue https://github.com/ubiquity/ubiquibot/issues/378

And this is a related PR https://github.com/ubiquity/ubiquibot/pull/377 (which had been reverted because we got rid of `mdast` packages)

Reply:
it seems https://github.com/ubiquity/ubiquibot/pull/377 was supposed to fix the issue but it doesn't - instead of throwing the error at startup, it throws the error when the function runs.

Reply:
So it throws the same error both when modules are imported at the top of the `comment.ts` file and in the `parseComments()` function, right?

Reply:
> So it throws the same error both when modules are imported at the top of the `comment.ts` file and in the `parseComments()` function, right?

yes

Reply:
I don't remember the whole context. As far as I remember this error occurs only during the build step, right?

Reply:
> As far as I remember this error occurs only during the build step, right?

It's runtime error. If it's imported at the top it throws error when you start the bot, if it's imported in the function it throws when the function runs.

Reply:
As far as I understand [this module](https://github.com/syntax-tree/mdast-util-from-markdown) is ESM only which means that we can use only `import` statements in the bot's code. Since we use commonjs in the bot's code then all `import()` statements are transpiled into `require()` statements hence the `Error [ERR_REQUIRE_ESM]: require() of ES Module` error.

I can only point you to the [docs](https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c). Seems like we need to modify smth in typescript settings. This [PR](https://github.com/ubiquity/ubiquibot/pull/377) solved the build step but I don't remember running any `mdast` related code in runtime so chances are that the code in that PR is also broken.

Reply:
I tried changing Typescript settings to transpile to ESM and added `"type": "module"` to `package.json`, but I get a different error because Probot is using `require()` internally

```
Error [ERR_REQUIRE_ESM]: require() of ES Module /Users/whilefoo/ubiquity/ubiquibot/lib/src/index.js from /Users/whilefoo/ubiquity/ubiquibot/node_modules/probot/lib/helpers/resolve-app-function.js not supported.
Instead change the require of index.js in /Users/whilefoo/ubiquity/ubiquibot/node_modules/probot/lib/helpers/resolve-app-function.js to a dynamic import() which is available in all CommonJS modules.
```

`node_modules/probot/lib/helpers/resolve-app-function.js` is using `require()`

Reply:
Perhaps we could downgrade all 3 packages (`mdast-util-from-markdown`, `mdast-util-gfm`, `micromark-extension-gfm`) to their "pre ESM" versions

Reply:
I tried downgrading and I have an issue with `mdast-util-gfm` because that version didn't have any types so typescript is complaining. I will see if there's anything I can do

Reply:
https://github.com/ubiquity/ubiquibot/blob/incentive-based-on-comment/src/helpers/comment.ts#L23

Please take a look at the `incentive-based-on-comment` branch @whilefoo. It has been done with `mdast` library. 
The blocker on that branch was the function timeout: 10s as of now.  

I hope you can get reusable/helpful methods from that branch.

Reply:
> Please take a look at the `incentive-based-on-comment` branch @whilefoo.

That is exactly how I've done it but it's not working, you can check the whole conversation for context